Add kernel module auto-loading

behlendorf opened the pull request:
### Description

Historically a dynamic misc minor number was registered for the
/dev/zfs device in order to prevent minor number collisions.  This
was fine but it prevented us from being able to use the kernel
module auto-loaded which requires a known reserved value.

Resolve this issue by adding a configure test to find an available
misc minor number which can then be used in MODULE_ALIAS_MISCDEV at
build time.  By adding this alias the zfs kmod is added to the list
of known static-nodes and the systemd-tmpfiles-setup-dev service
will create a /dev/zfs character device at boot time.

This in turn allows us to update the 90-zfs.rules file to make it
aware this is a static node.  The upshot of this is that whenever
a process (zpool, zfs, zed) opens the /dev/zfs the kmods will be
automatic loaded.  This even works for unprivileged users so there
is no longer a need to manually load the modules at boot time.

As an additional bonus the zed now no longer needs to start after
the zfs-import.service since it will trigger the module load.

In the unlikely event the minor number we selected conflicts with
another out of tree unregistered minor number the code falls back
to dynamically allocating it.  In this case the modules again
must be manually loaded.

### Motivation and Context

This is a long overdue bit of functionality.  This way things will just work.

### How Has This Been Tested?

Manually on Fedora 27.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

Fabian-Gruenbichler commented on config/kernel-misc-minor.m4:
```diff
@@ -0,0 +1,26 @@
+	AC_MSG_CHECKING([for available /dev/zfs minor])
+
+	for i in $(seq 249 -1 200); do
+		if ! grep -q "#define.*_MINOR.*$i" \
```
`"^#define\s\+.*_MINOR\s\+.*$i"` would be a bit more exact and prevent potential future false negatives (they currently match the same device minor numbers).

Fabian-Gruenbichler commented on module/zfs/zfs_ioctl.c:
```diff
@@ -6962,12 +6965,17 @@ zfs_attach(void)
+		 * In this case the kernel modules must be manually loaded.
+		 */
+		zfs_misc.minor = MISC_DYNAMIC_MINOR;
+		error = misc_register(&zfs_misc);
```
keeping the existing printk (moved into a second if after the EBUSY one) and adding a second printk inside the if like this:
```
printk(KERN_INFO "ZFS: misc_register() with static minor %d failed %d, retrying with MISC_DYNAMIC_MINOR\n", ZFS_MINOR, error);
```

would make debugging issues here a lot easier I think (although they will probably trigger rarely if at all).

utopiabound commented:
So, if I understand the udev rules.  udev will load zfs modules if any block devices are part of a zpool.  But on initial installation, we still have to worry about loading module prior to zed starting?

jgrund commented:
Will this load the kernel-module (and start dependent services) when installing ZFS for the first time?

behlendorf commented:
@Fabian-Gruenbichler thanks for the review and keeping me honest!  I've updated the PR and addressed your review feedback.

I also did a little testing of the patch on CentOS 7 and discovered a few small issues which have been subsequently fixed upstream in Fedora.

https://lists.freedesktop.org/archives/systemd-devel/2014-October/024673.html

Anyway the take away is that for CentOS 7 for things to just work after installing the packages (without a reboot) we need to manually restart the `kmod-static-nodes` and `systemd-tmpfiles-setup-dev` services, then run `udevadm trigger` to properly permission the `/dev/zfs` device.  For simplicity I've opted to do this in the `%posttrans` section and only for CentOS 7.  This does mean you'll want to use `yum/dnf` to install the packages instead of directly calling `rpm`.

@utopiabound @jgrund it would be great if you could verify this PR does work as intended on your sytstems.

> So, if I understand the udev rules. udev will load zfs modules if any block devices are part of a zpool.

Yes.  It will proactively load the modules for you, although that really shouldn't been needed anymore with this change.

>  But on initial installation, we still have to worry about loading module prior to zed starting?

Nope, with this change starting the zed will trigger the module load.  Effectively the zed service has been completely decoupled from the import step and it can run at any time.  It is still the case that in accordance with the Fdora packaging guidelines the service won't be automatically started as part of the install, only enabled.

```
$ yum install zfs
$ systemctl start zfs-zed
$ systemctl status zfs-zed
● zfs-zed.service - ZFS Event Daemon (zed)
   Loaded: loaded (/usr/lib/systemd/system/zfs-zed.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2018-03-09 20:52:07 UTC; 49s ago
     Docs: man:zed(8)
 Main PID: 1428 (zed)
   CGroup: /system.slice/zfs-zed.service
           └─1428 /sbin/zed -F

Mar 09 20:52:07 ip-172-31-21-231.us-west-1.compute.internal systemd[1]: Started ZFS Event Daemon (zed).
Mar 09 20:52:07 ip-172-31-21-231.us-west-1.compute.internal systemd[1]: Starting ZFS Event Daemon (zed)...
Mar 09 20:52:07 ip-172-31-21-231.us-west-1.compute.internal zed[1428]: ZFS Event Daemon 0.7.0-361_gdba361e (PID 1428)
Mar 09 20:52:08 ip-172-31-21-231.us-west-1.compute.internal zed[1428]: Processing events since eid=0
...

jgrund commented:
@behlendorf What release with this change be part of?

behlendorf commented:
@jgrund it will be part of the 0.8 release, and is under consideration for inclusion in a future 0.7.x point release.

jgrund commented:
Thanks

kpande commented:
this breaks importing root pool on boot in dracut, guys. was that tested?

behlendorf commented:
@kpande sorry!  That case didn't get covered in the manual testing.  I would have expected the updated udev rule to take care of it, but it sounds like that isn't the case.  Did you settle on a solution?

kpande commented:
nope, haven't had time to get into it, i've had to mount manually each time I boot.

behlendorf commented:
@Mic92 it may be needed early in the boot before the normal kernel module autoloading infrastructure is in place.

qiaoqiang commented:
When no pool has been created at all，and all disks are free to use which have no metadata about zfs_member,  manually modprobing zfs is required after reboot. I found this situation in CentOS 7.2.
